Library/Homebrew/test/formula_info_spec.rb

```diff
@@ -5,6 +5,8 @@
 
 describe FormulaInfo, :integration_test do
   it "tests the FormulaInfo class" do
+    ENV["HOMEBREW_DEVELOPER"] = "1"
```

Comment:
Why is this change needed?

Reply:
These changes are needed because the following tests will call `brew` in sub shell.

In my test with portable ruby, the correct ruby can only be picked up if both `HOMEBREW_DEVELOPER` and `HOMEBREW_RUBY_PATH` are set.

So I manually set `HOMEBREW_DEVELOPER` here.

Reply:
https://github.com/Homebrew/brew/blob/master/Library/Homebrew/utils/ruby.sh#L14-L17 is what unsets `HOMEBREW_RUBY_PATH` if `HOMEBREW_DEVELOPER` isn't set. As a result, `HOMEBREW_DEVELOPER` should only be needed in cases where `/usr/local/bin/brew` is called rather than `HOMEBREW_LIBRARY_PATH/"brew.rb"` (as with the integration tests). If it's only needed in this single integration test then the `FormulaInfo` class should be modified to run the same way as the integration tests are being run.

Reply:
The issue is that `HOMEBREW_BREW_FILE` is called as a sub process in normal brew code (outside test case)
 
For example: 
https://github.com/Homebrew/brew/blob/1798deda057c409cc13b9e85a6f88419713db0a3/Library/Homebrew/formula_info.rb#L16

Reply:
Yup, I understand and agree that should be probably changed to instead use `brew.rb` instead. To be explicit, though: this isn't a fix for ARM, it's a fix for running `brew tests` with a custom `HOMEBREW_RUBY_PATH`.

This doesn't explain why the other changes are needed.

Reply:
> though: this isn't a fix for ARM, it's a fix for running brew tests with a custom HOMEBREW_RUBY_PATH.

Yes. I just put this fix with other fixes in the same PR.

> This doesn't explain why the other changes are needed.

I have removed the changes on `integration_test.rb`. They should not be necessary.

Reply:
> Yes. I just put this fix with other fixes in the same PR.

Ok. Could you try to fill out the PR description in future? It makes reviewing easier.

Reply:
I'd like to avoid setting `HOMEBREW_DEVELOPER` here. Either refactor `formula_info` to call with `brew.rb` (as integration tests do) or make this assignment conditional on a custom Ruby path somehow.

Reply:
> I'd like to avoid setting HOMEBREW_DEVELOPER here.

Why? This only affects one test case. 

>   Either refactor formula_info to call with brew.rb (as integration tests do) 

I don't think the following code in `formula_info` is easy to read comparing with current one. Noted that they are not test code.
```ruby
json = Utils.popen_read(
      RUBY_PATH,
      "-W0",
      "-I", $LOAD_PATH.join(File::PATH_SEPARATOR),
      HOMEBREW_LIBRARY_PATH/"brew.rb",
      "info",
      "--json=v1",
      name
    )
```

> make this assignment conditional on a custom Ruby path somehow.

There is no way to tell different between a custom ruby path here and normal ruby from `PATH`. Noted that in Linux, using Ruby from system is normal behaviour. Further, it makes no sense to make test behaviour different because different ruby are used.

Reply:
If you really want to make code more clean, we should change https://github.com/Homebrew/brew/blob/master/Library/Homebrew/utils/ruby.sh#L14-L17 to use a different flag for testing purpose. 

For example,
```bash
  if [[ -z "$HOMEBREW_TEST_RUBY_PATH" ]]
  then
    unset HOMEBREW_RUBY_PATH
  fi
```

Reply:
> Why? This only affects one test case.

It's a hack that's only necessary with a custom `HOMEBREW_RUBY_PATH` and `HOMEBREW_DEVELOPER` often changes the behaviour of commands. A custom `HOMEBREW_RUBY_PATH` is not something we support for non-developers.

> I don't think the following code in `formula_info` is easy to read comparing with current one. Noted that they are not test code.

It's not as easy but it is more correct. We could provide a helper method if abstraction would be better.

Reply:
I think my approach in https://github.com/Homebrew/brew/pull/6617#discussion_r336498383 should avoid problems you mentioned while still keeping the code easy to read.

Reply:
That's adding another undocumented conditional for an edge-case only needed for running tests in an unsupported configuration. I'm sorry, I disagree.

Reply:
OK, updated PR as you suggest.